x-pack/legacy/plugins/security/server/lib/authentication/tokens.ts

```diff
@@ -0,0 +1,173 @@
+    try {
+      // Token should be refreshed by the same user that obtained that token.
+      const {
+        access_token: accessToken,
+        refresh_token: refreshToken,
+      } = await this.options.client.callWithInternalUser('shield.getAccessToken', {
+        body: { grant_type: 'refresh_token', refresh_token: existingRefreshToken },
+      });
+
+      this.debug('Access token has been successfully refreshed.');
+
+      return { accessToken, refreshToken };
+    } catch (err) {
+      this.debug(`Failed to refresh access token: ${err.message}`);
+
+      // There are at least two common cases when refresh token request can fail:
+      // 1. Refresh token is valid only for 24 hours and if it hasn't been used it expires.
+      //
+      // 2. Refresh token is one-time use token and if it has been used already, it is treated in the same way as
+      // expired token. Even though it's an edge case, there are several perfectly valid scenarios when it can
+      // happen. E.g. when several simultaneous AJAX request has been sent to Kibana, but access token has expired
+      // already, so the first request that reaches Kibana uses refresh token to get a new access token, but the
+      // second concurrent request has no idea about that and tries to refresh access token as well. All ends well
+      // when first request refreshes access token and updates session cookie with fresh access/refresh token pair.
+      // But if user navigates to another page _before_ AJAX request (the one that triggered token refresh) responds
+      // with updated cookie, then user will have only that old cookie with expired access token and refresh token
+      // that has been used already.
+      //
+      // When user has neither valid access nor refresh token, the only way to resolve this issue is to get new
+      // SAML LoginResponse and exchange it for a new access/refresh token pair. To do that we initiate a new SAML
+      // handshake. Obviously we can't do that for AJAX requests, so we just reply with `400` and clear error message.
+      // There are two reasons for `400` and not `401`: Elasticsearch search responds with `400` so it seems logical
+      // to do the same on Kibana side and `401` would force user to logout and do full SLO if it's supported.
+      if (getErrorStatusCode(err) === 400) {
+        this.debug('Refresh token is either expired or already used.');
+        return null;
+      }
+
+      throw err;
+    }
+  }
+
+  /**
+   * Tries to invalidate provided access and refresh token pair. At least one of the tokens should
+   * be specified.
+   * @param [accessToken] Optional access token to invalidate.
+   * @param [refreshToken] Optional refresh token to invalidate.
+   */
+  public async invalidate({ accessToken, refreshToken }: Partial<TokenPair>) {
+    this.debug('Invalidating access/refresh token pair.');
+
+    let invalidationError;
+    let invalidatedTokensCount;
+    if (refreshToken) {
```

Comment:
__note:__ I know it's kind of stupid since invalidation will be super quick, but I feel bit safer if I invalidate refresh token first as having it active feels more dangerous :see_no_evil: 

Another change in behavior here is that I'll try to invalidate access token even if invalidation of refresh token failed previously.